README.md
# devchain

Private Ethereum blockchain creator and development geth server

With **devchain** you can easily create your own private Ethereum blockchains and private networks. The setup needed is abstracted away, and you can create multiple private blockchains in seconds!

devchain is great for development because you can adjust mining difficulty and keep your blockchain state over multiple sessions,  Plus, since you're using a real geth node, production release on geth will be a breeze! You can also quickly set up your own private networks to test on.

Additionally, a Javascript file is preloaded into geth that gives you a ton of useful features. It automates account creation, distributing Ether to your accounts, and mining for transactions. It also displays transaction info and contains an object (`dev`) containing helper methods such as showing account balances and transferring Ether between your accounts.

## Install
You must [install geth](https://github.com/ethereum/go-ethereum/wiki/Building-Ethereum) to use devchain.

After that install devchain:

```
npm install -g devchain
```

## Usage

### Initialize
In your terminal type:

```
-> devchain
```
The first time you use the command it will initialize devchain.

In your current directory this will create a folder called `devchain/`, a file called `devgenesis.json`, and a file called `devconfig.js`. The folder will contain your Ethereum blockchain data such as blocks and accounts. The file `devgenesis.json` is your blockchain's genesis file, which contains the data for the blockchain's first block. [Go here](http://ethereum.stackexchange.com/questions/2376/what-does-each-genesis-json-parameter-mean) to learn more about the genesis file.

The genesis file's most useful property is the "difficulty". The difficulty refers to how fast new blocks can be mined. You can set a low difficulty for development purposes or a high difficulty to mimic actual production.

The file `devconfig.js` lets you adjust the options of the geth preload script and the geth server options. The default options are shown below. All [command options](#options) will overwrite the config file's expect staticNodes, which gets added on instead.

```
{
  /** Preload script options */
  autoMine: true, // Set false to turn off auto mining
  isMute: false, // Set true to turn off transaction receipt display
  accountAmount: 3, // Number of accounts to create
  password: '', // Password to create accounts with
  minAmount: 50, // Ether amount to keep coinbank topped off at
  distributeAmount: 10, // Ether amount to distribute to all accounts
  blocks: 5, // Additional blocks to mine after auto mining a transaction

  /** Custom geth node start options */
  rpcaddr: 'localhost', // RPC host
  rpcport: 8545, // RPC port
  port: 30303, // P2P network listening port
  networkid: 1, // Network identifier. To connect with other nodes
  staticNodes: [] // Geth enode addresses to connect with
}
```

Adjust `devgenesis.json` and `devconfig.js` as needed.

### Start
Call the devchain command again to start up the geth server. If there's no blockchain data it will initialize a new blockchain based on `devgenesis.json`.

If you wish to have a set of accounts all preallocated with Ether wait until the initial setup actions are completed. Setup actions can be adjusted from `devconfig.js` or thru [command options](#options). If you set your own password you will need to leave it inside the config file or remember it for next time to pass in as an option.

By default the geth server will create 3 accounts and automatically start mining. 10 Ether will be distributed from your coinbase (first account created) to all other accounts after 31 Ether is reached. It will then mine 50 Ether and wait for pending transactions before mining again, or to keep the coinbase topped off.

Ether will only be distributed to all accounts on a new blockchain geth start. To disable this feature turn off auto mining with the option `--off`. You can turn auto mining back on with the geth console's `dev` object.

### Console helper methods
In the geth console you're given access to an object called `dev`. The methods are shown below. These methods are also shown upon starting up the geth server or by using `dev.help()`.

* **dev**
   * **.minAmount** - *Adjust the minimum amount of Ether to keep above*
   * **.accounts()** - *Displays all accounts, balances, and indexes*
   * **.auto()** - *Toggles auto mining*
   * **.start(threads)** - *Start mining -- threads defaults to 1*
   * **.stop()** - *Stop mining*
   * **.transfer(fromIndex, toIndex, etherAmount)** - *Transfer Ether between your accounts*
   * **.distribute(fromIndex, etherAmount)** - *Distribute Ether to all your accounts from one account*
   * **.mine(blockAmount)** - *Mine a certain amount of blocks -- blockAmount defaults to 1*
   * **.block(blockNumber)** - *Display block information -- blockNumber defaults to latest*
   * **.coinbase(accountIndex)** - *Change coinbase*
   * **.mute()** - *Toggles transaction receipt display*
   * **.help()** - *Display dev methods*

### Connection
The geth option to enable an RPC connection, `--rpc`, is used by default. The connection listens at the address and port that you set. The default is `localhost:8545`. To turn off the RPC connection use the option `--rpcoff`.

### Reset
To reset your blockchain data and accounts use:

```
-> devchain --reset
```

This will create a new `devchain/` folder and if the `devgenesis.json` exists in the current directory it will start up a new blockchain.

### Connecting with other blockchains
To connect with other blockchains you will need the geth enode addresses you wish to connect with. The other blockchains will need to have the same genesis file and be using the same network id as you. Your enode address is shown when you start the geth server.

You can pass in static nodes with the command option `--staticnodes` or you can add them into the `devconfig.js` staticNodes property. For new blockchains it creates a `static-nodes.json` file within your blockchain folder, otherwise it connects them within geth. [Go here](https://github.com/ethereum/go-ethereum/wiki/Connecting-to-the-network) to learn more about static nodes and connecting to peers.

The geth option `--nodiscover` is used by default so only people who manually add you will be able to find you. Use the option `--nodiscoveroff` to disable it.

To connect with nodes running on the network id 1000:

```
-> devchain --staticnodes enode://pubkey1@ip:port,enode://pubkey2@ip:port --networkid 1000
```

You can create multiple private blockchains running on your computer by giving them an unique RPC port, network p2p port, and blockchain data location.

The default ports opened:
```
-> devchain --rpcport 8545 --port 30303
```

Or with the shorthand options:
```
-> devchain -p 8545 -n 30303
```

You can specify a custom path to the blockchain data. Alternatively you can go to that directory to call the command.

```
-> devchain --datadir './relative/path/to/chaindata' --rpcport 8546 --port 30304
```

<a name="options"></a>
## Options
All command options will overwrite the `devconfig.js` options except `--staticnodes`, which gets added on instead.

**devchain `-r --reset, --off, --mute, -a --accounts <number>, -t --distribute <number>, -w --password <value>, -b --blocks <number>, -d --datadir <path>, -h --rpcaddr <value>, -p --rpcport <number>, -n --port <number>, --networkid <number>, --rpcoff, --nodiscoveroff, -s --staticnodes <enodes>`**

```
Options             Type          Description

-r --reset           --           Reset blockchain data
--off                --           Turn off auto mine
--mute               --           Turn off transaction receipt display
-a --accounts        <number>     Number of accounts to initially create. Default is 3
-t --distribute      <number>     Ether amount to distribute to accounts initially. Default is 10
-w --password        <value>      Password to give and unlock the accounts created. Default is ""
-b --blocks          <number>     Additional blocks to mine after auto mining a transaction
-d --datadir         <path>       Relative path to blockchain data
-h --rpcaddr         <value>      Geth server HTTP-RPC addr. Default is 'localhost'
-p --rpcport         <number>     Geth server HTTP-RPC port. Default is 8545
-n --port            <number>     Geth server network p2p port. Default is 30303
--networkid          <number>     Geth network identifier. To connect with other nodes. Default is 1
--rpcoff             --           Disable the Geth RPC connection. Don't set `--rpc` option for geth
--nodiscoveroff      --           Don't set `--nodiscover` option for geth
-s --staticnodes     <enodes>     Comma separated list with no spaces of static nodes to connect with.
                                  This will create a static-nodes.json file within your blockchain's data directory for new blockchains
```

You can pass in additional options that aren't listed for geth. You can also do this via `devconfig.js` as additional properties. The key will be the option name with the '--' removed. If your desired option doesn't require a value, set it as true. An example is shown below:

```
{
  ipcdisable: true,
  ipcapi: 'admin, eth, personal, web3',
  fast: true
}
```


CHANGELOG.md
# Changelog

## 1.2 (11-26-16)

* Added shorthand options for `--rpcport` and `--port`. These are `-r` and `-n` respectively.
* Added additional blocks to mine after auto mining a transaction to options and config. `--blocks`
* Min amount gets adjusted if its less than accounts * distributeAmount.


src/cli/extraArgs.js
'use strict';
const defaultConfig = require('./../devchain/devconfig');

/**
 * Filters args for the properties in a commander object
 * @param {Array} args - An array of args to filter
 * @param {Object} commandOptions - Object of commander options
 * @param {Object} config - Config object
 * @returns {Array} - Array of filtered args
 */
module.exports = (args, commandOptions, config) => {
  let filter = true;
  const newArgs = [];

  for (let i = 0; i < args.length; i++) {
    if (args[i].match(/^--/) || args[i].match(/^-/)) {
      if (!commandOptions.hasOwnProperty(args[i])) {
        newArgs.push(args[i]);
        filter = false;
      } else {
        filter = true;
      }
    } else if (!filter && typeof args[i] === 'string' || typeof args[i] === 'number') {
      newArgs.push(args[i]);
    }
  }

  /** Add additional config properties to rawArgs */
  for (let key in config) {
    if (!defaultConfig.hasOwnProperty(key) && !commandOptions.hasOwnProperty('--' + key)) {
      newArgs.push('--' + key);
      if (typeof config[key] === 'string' || typeof config[key] === 'number') {
        newArgs.push(config[key]);
      }
    }
  }

  return newArgs;
};


src/cli/createConfig.js
'use strict';
const path = require('path');
const pathExists = require('path-exists').sync;
const defaultConfig = require('./../devchain/devconfig');

/**
 * Create a config object based on commander options
 * @param {string} configPath - Path to config file
 * @param {Command} options - Commander object with process.argv parsed
 * @returns {Object} Config object
 */
module.exports = (configPath, options) => {
  /** Check to see if devconfig.js is there to overwrite default config options */
  let config = {};
  if (pathExists(configPath)) {
    config = require(path.relative(__dirname, configPath));
    config = Object.assign({}, defaultConfig, config);
  } else {
    config = Object.assign({}, defaultConfig);
  }

  /** Set config based on options */
  config.autoMine = (options.off || config.autoMine === false) ? false : true;
  config.accountAmount = options.accounts || config.accountAmount;
  config.password = options.password || config.password;
  config.distributeAmount = options.distribute || config.distributeAmount;
  config.isMute = (options.isMute || config.isMute === true) ? true : false;
  config.rpcaddr = options.rpcaddr || config.rpcaddr;
  config.rpcport = options.rpcport || config.rpcport;
  config.port = options.port || config.port;
  config.networkid = options.networkid || config.networkid;

  if (options.staticnodes) {
    options.staticnodes = options.staticnodes.split(',')
      .map(node => node.trim());
    config.staticNodes = config.staticNodes.concat(options.staticnodes);
  }

  return config;
};


src/devchain/devconfig.js
/** Config file for devchain. This file is not required to use devchain */
module.exports = {

  /** Preload script options */
  autoMine: true, // Set false to turn off auto mining
  isMute: false, // Set true to turn off transaction receipt display
  accountAmount: 3, // Number of accounts to create
  password: '', // Password to create accounts with
  minAmount: 50, // Ether amount to keep coinbank topped off at
  distributeAmount: 10, // Ether amount to distribute to all accounts
  blocks: 5, // Additional blocks to mine after auto mining a transaction

  /** Custom geth node start options */
  rpcaddr: 'localhost', // RPC address
  rpcport: 8545, // RPC port
  port: 30303, // P2P network listening port
  networkid: 1, // Network identifier. To connect with other nodes
  staticNodes: [], // Geth enode addresses to connect with
};


src/devchain/devchain.js
var CONFIG = {"autoMine":true,"isMute":false,"accountAmount":3,"password":"","minAmount":50,"distributeAmount":10,"rpcaddr":"localhost","rpcport":8545,"port":30303,"networkid":1,"staticNodes":[],"path":"/Users/zhiwen/Development/devchain/test/devchain"}; 
/* global web3, CONFIG */

/** The global dev methods */
var dev = CONFIG; // CONFIG object gets built onto this file

/** Displays all accounts, balances, and indexes */
dev.accounts = function() {
  console.log('');
  this.log('=== Accounts ===');
  this.log('Index  Account                                      Ether');
  for (var i = 0; i < web3.eth.accounts.length; i++) {
    if (web3.eth.accounts[i] === web3.eth.coinbase) {
      dev.log(i, '    ', web3.eth.accounts[i], ' ', this.weiToEther(web3.eth.getBalance(web3.eth.accounts[i]).toString(10)), '- coinbase');
    } else {
      dev.log(i, '    ', web3.eth.accounts[i], ' ', this.weiToEther(web3.eth.getBalance(web3.eth.accounts[i]).toString(10)));
    }
  }
  console.log('');
  return true;
};

/**
 * To start mining
 * @param {number} threads - Optional. The number of threads to mine with. Defaults to 1/
 */
dev.start = function(threads) {
  threads = threads || 1;
  if (!web3.eth.mining) {
    console.log('');
    dev.log('Starting miner');
    console.log('');
    web3.miner.start(threads);
    return true;
  }
  return false;
};

/** To stop mining */
dev.stop = function() {
  if (web3.eth.mining) {
    console.log('');
    dev.log('Stopping miner');
    console.log('');
    web3.miner.stop();
    return true;
  }
  return false;
};

/** Toggles auto mining */
dev.auto = function() {
  if (this.autoMine === true) {
    this.autoMine = false;
    console.log('');
    this.log('Auto mining is off');
    console.log('');
  } else {
    this.autoMine = true;
    console.log('');
    this.log('Auto mining is on');
    console.log('');
  }
  this.stop();
  return true;
};


/**
 * Distributes Ether to all of your accounts from an account
 * @param {number} fromIndex - Index in web3.eth.accounts to sender Ether from
 * @param {number} etherAmount - Amount to transfer to all accounts
 */
dev.distribute = function(fromIndex, etherAmount) {
  var requiredAmount = etherAmount * (web3.eth.accounts.length - 1) + 2;
  if (web3.eth.getBalance(web3.eth.accounts[fromIndex]).greaterThan(this.etherToWei(requiredAmount))) {
    for (var i = 0; i < web3.eth.accounts.length; i++) {
      if (i === fromIndex) continue;
      var object = {
        from: web3.eth.accounts[fromIndex],
        to: web3.eth.accounts[i],
        value: this.etherToWei(etherAmount)
      };
      web3.eth.sendTransaction(object);
    }
    console.log('');
    this.log(etherAmount, 'Ether distributed to accounts');
    console.log('');
    this.mine(2);
    return true;
  } else {
    var balanceEther = this.etherToWei(web3.eth.getBalance(web3.eth.accounts[fromIndex]));
    console.log('');
    this.log('Not enough Ether. Balance:', balanceEther, 'Required:', requiredAmount);
    console.log('');
    return false;
  }
};

/**
 * Transfer Ether between your accounts
 * @param {number} fromIndex - Index in web3.eth.accounts to send Ether from.
 * @param {number} toIndex - Index in web3.eth.accounts to send Ether to.
 * @param {number} etherAmount - Amount of Ether to transfer over
 */
dev.transfer = function(fromIndex , toIndex, etherAmount) {
  if (fromIndex === toIndex) return;
  if (fromIndex < 0 || toIndex < 0) return;
  if (fromIndex > web3.eth.accounts.length - 1 || toIndex > web3.eth.accounts.length - 1) return;

  var balanceEther = this.weiToEther(web3.eth.getBalance(web3.eth.accounts[fromIndex]));
  if (balanceEther < etherAmount) {
    console.log('');
    this.log('Not enough Ether. Balance:', balanceEther);
    console.log('');
    return;
  }
  var object = {
    from: web3.eth.accounts[fromIndex],
    to: web3.eth.accounts[toIndex],
    value: this.etherToWei(etherAmount)
  };
  web3.eth.sendTransaction(object);
  console.log('');
  this.log(etherAmount, 'Ether transfered to', web3.eth.accounts[fromIndex], 'from', web3.eth.accounts[toIndex]);
  console.log('');
  this.mine(2);
  return true;
};

/**
 * Mine a specified number of blocks
 * @param {number} blockAmount - Number of blocks to mine
 */
dev.mine = function(blockAmount) {
  blockAmount = blockAmount || 1;
  if (blockAmount <= 0 || web3.eth.mining === true) return false;
  var stopBlock = web3.eth.blockNumber + blockAmount;
  this.start();

  // keep ref to auto mine status when called to set it back to it later
  var autoStatus = this.autoMine;
  this.autoMine = false;

  // Create the filter to watch with check function
  var filter = web3.eth.filter('latest');
  filter.watch(check.bind(this));

  console.log('');
  dev.log('Mining', blockAmount, 'blocks');
  console.log('');

  function check() {
    if (web3.eth.blockNumber >= stopBlock || web3.eth.mining === false) {
      // Stop watching filter
      this.autoMine = autoStatus;
      if (this.autoMine !== true) this.stop();
      filter.stopWatching();
    }
  }
  return true;
};

/**
 * Display block info
 * @param {number} blockNumber - Optional. The block number to display the info of. Defaults to 'latest'.
 */
dev.block = function(blockNumber) {
  blockNumber = blockNumber || 'latest';
  web3.eth.getBlock(blockNumber, function(err, block) {
    if (!err) {
      console.log('');
      this.log('Block Info');
      this.log('Number:     ' + block.number);
      this.log('Hash:       ' + block.hash);
      this.log('Difficulty: ' + block.difficulty);
      this.log('Gas Limit:  ' + block.gasLimit.toString());
      this.log('Gas Price:  ' + web3.eth.gasPrice.toString());
      this.log('Total Gas:  ' + block.gasUsed);
      this.log('Miner:      ' + block.miner);
      console.log('');
    }
  }.bind(this));
  return true;
};

/**
* Change the coinbase to another one of your accounts. This is where the Ether mining rewards are deposited.
* @param accountIndex - Index in web3.eth.accounts.
*/
dev.coinbase = function(accountIndex) {
  if (accountIndex < 0 || accountIndex > web3.eth.accounts.length - 1) return false;
  web3.miner.setEtherbase(web3.eth.accounts[accountIndex]);
  console.log('');
  this.log('Coinbase set to', web3.eth.coinbase);
  console.log('');
  return true;
};

/** Wei to Ether conversion */
dev.weiToEther = function (amount) {
  return Number(web3.fromWei(amount, 'ether').toString());
};

/** Ether to Wei conversion */
dev.etherToWei = function(amount) {
  return Number(web3.toWei(amount, 'ether').toString());
};

/** dev console log */
dev.log = function() {
  var args = Array.prototype.slice.call(arguments);
  args.unshift('[dev]');
  args.concat(arguments);
  console.log.apply(this, args);
};

/** Toggles transaction receipt display status */
dev.mute = function() {
  this.isMute = this.isMute ? false : true;
  if (this.isMute === true) {
    console.log('');
    this.log('Transaction receipts are muted');
    console.log('');
  } else {
    console.log('');
    this.log('Transaction receipts are unmuted');
    console.log('');
  }

  return true;
};

/** Display dev methods */
dev.help = function() {
  console.log('');
  this.log('=== devchain Methods ===');
  this.log('dev');
  this.log(' .minAmount', '                           Set the minimum amount to mine to');
  this.log(' .accounts()', '                          Displays all accounts, balances, and indexes');
  this.log(' .auto()', '                              Toggles auto mining');
  this.log(' .start(threads)', '                      Start mining', '- threads defaults to 1');
  this.log(' .stop()', '                              Stop mining');
  this.log(' .transfer(fromIndex, toIndex, ether)', ' Transfer Ether between your accounts');
  this.log(' .distribute(fromIndex, ether)   ', '     Distribute Ether to all of your accounts from an account');
  this.log(' .mine(blockAmount)', '                   Mine a specified number of blocks', '- blockAmount defaults to 1');
  this.log(' .block(blockNumber)', '                  Display block info', '- blockNumber defaults to latest');
  this.log(' .coinbase(accountIndex)', '              Change coinbase');
  this.log(' .mute()', '                              Toggles transaction receipt display');
  this.log(' .help()', '                              Display devchain methods');
  console.log('');
  return true;
};

/**
 * Run on devchain start
 */
(function () {
  // Amount of Ether for coinbank to always have
  dev.minAmount;
  // Transaction receipt display status
  dev.isMute;
  // Amount of accounts to create. Always creates 1 account.
  var accountAmount = (dev.accountAmount > 0) ? dev.accountAmount : 1;
  accountAmount -= web3.eth.accounts.length; // Don't allow creation if accounts are already there
  // Amount of Ether to distribute to accounts
  var distributeAmount = (dev.distributeAmount < 0) ? 0 : dev.distributeAmount;
  // Amount of Ether needed by coinbank
  var requiredEtherAmount = dev.distributeAmount * dev.accountAmount + 1;
  // Status of toggling mining if there are transactions pending and whether to keep coinbank topped off at minAmount
  var password = dev.password; // password for the accounts created
  var blockNumber = web3.eth.blockNumber; // the block number to keep mining too

  // For the automatic Ether distribution
  var distributeBlock; // block number when distributed for reference
  var isBalanceDisplayed = false;
  var isDistributed = (distributeAmount === 0 || !dev.reset) ? true : false; // status of ether distribution

  var transactions = []; // to contain pending transaction hashes of a block
  var pendingBlock; // to make sure transactions are only displayed after another block is mined
  if (!dev.reset) {
    addPeers(dev.staticNodes);
  }

  console.log('');
  dev.log('=== Devchain Development Server ===');
  console.log('');
  dev.log('Path to blockchain data:', dev.path);
  console.log('');
  dev.log('Node address:', web3.admin.nodeInfo.enode);
  dev.help(); // Display dev methods and info

  if (dev.autoMine === true) {
    dev.log('Auto mining is on');
    dev.log('Coinbase will be remain above', dev.minAmount, 'Ether. Set dev.minAmount to adjust');
    dev.log('Mining will start automatically when there are transactions pending');
    if (distributeAmount !== 0) {
      dev.log('Distributing', distributeAmount, 'to accounts when coinbase has', requiredEtherAmount, 'Ether');
    }
  } else {
    isDistributed = true; // Don't allow for auto distribution if auto is off at the start
    dev.log('Auto mining is off');
    dev.log('To start/stop mining call dev.start() / dev.stop()');
    dev.log('To toggle auto mining call dev.auto()');
  }

  // Creates all accounts
  if (accountAmount > 0) {
    createAccounts(accountAmount, password);
  } else {
    unlockAccounts();
  }
  dev.accounts();

  // To allow for continous checking of status
  web3.eth.filter('latest', checkStatus);
  web3.eth.filter('pending', checkStatus);

  if (dev.autoMine === true) dev.start();

  function checkStatus() {
    // Display the transaction receipt of previous blocks
    if (transactions.length > 0 && web3.eth.blockNumber > pendingBlock) {
      for (var i = 0; i < transactions.length; i++) {
        var transactionHash = transactions[i];
        transactionReceipt(transactionHash);
      }
      transactions = [];
    }

    // Fill transactions array to display receipts in next bock
    if (web3.eth.getBlock('pending').transactions.length > 0) {
      pendingBlock = web3.eth.blockNumber;
      transactions = web3.eth.getBlock('pending').transactions;
    }

    // Auto mining actions
    if (dev.autoMine === true) {
      // For distributing Ether to accounts and displaying it
      if (isBalanceDisplayed === false && distributeBlock + 2 <= web3.eth.blockNumber) {
        dev.accounts();
        isBalanceDisplayed = true;
      }
      if (isDistributed === false) {
        distributeEther();
      }

      if (web3.eth.getBlock('pending').transactions.length > 0) {
        blockNumber = web3.eth.blockNumber + 3;
        dev.start();
        return;
      }

      if (web3.eth.getBalance(web3.eth.coinbase).lessThan(dev.etherToWei(dev.minAmount))) {
        dev.start();
        return;
      }

      if (web3.eth.blockNumber > blockNumber) {
        dev.stop();
        return;
      }
    }
  }

  // Create the specified number of accounts
  function createAccounts(accountAmount, password) {
    if (accountAmount <= 0) return;
    console.log('');
    dev.log('Creating ' + accountAmount + ' accounts with password ' + '"' + password + '"');
    for (var i = 0; i < accountAmount; i++) {
      web3.personal.newAccount(password);
      web3.personal.unlockAccount(web3.eth.accounts[web3.eth.accounts.length - 1], password, 10000000);
      dev.log('.');
    }
  }

  function unlockAccounts() {
    console.log('');
    dev.log('Unlocking accounts');
    for (var i = 0; i < web3.eth.accounts.length; i++) {
      try {
        web3.personal.unlockAccount(web3.eth.accounts[i], password, 10000000);
        dev.log('.');
      } catch (e) {
        dev.log('Unable to unlock account', i);
        dev.log('Unlock with: web3.personal.unlockAccount(' + web3.eth.accounts[i] + ', \'yourpassword\', 100000)');
      }
    }
    console.log('.');
  }
  // Display the transaction receipts
  function transactionReceipt(transactionHash) {
    if (dev.isMute === true) return;

    var transactionObj = web3.eth.getTransactionReceipt(transactionHash);
    console.log('');
    dev.log('=== Transaction Receipt ===');
    dev.log('Block:            ', transactionObj.blockNumber);
    dev.log('Hash:             ', transactionObj.transactionHash);
    dev.log('From:             ', transactionObj.from);
    dev.log('To:               ', transactionObj.to);
    dev.log('Gas Used:         ', transactionObj.gasUsed);
    dev.log('Ether Cost Est:   ', dev.weiToEther(web3.eth.gasPrice * transactionObj.gasUsed));
    dev.log('Created Contract: ', transactionObj.contractAddress);
    console.log('');
  }

  // Auto distribute ether to all your accounts
  function distributeEther() {
    if (web3.eth.getBalance(web3.eth.coinbase).greaterThan(dev.etherToWei(requiredEtherAmount))) {
      var doesOneDistribute = false; // check to see if there needs to be a blank line consoled
      for (var i = 0; i < web3.eth.accounts.length; i++) {
        if (web3.eth.accounts[i] !== web3.eth.coinbase && web3.eth.getBalance(web3.eth.accounts[i]) == 0) {
          var object = {
            from: web3.eth.coinbase,
            to: web3.eth.accounts[i],
            value: dev.etherToWei(distributeAmount)
          };
          web3.eth.sendTransaction(object);
          if (!doesOneDistribute) {
            console.log('');
            doesOneDistribute = true;
          }
          dev.log('Distributed', distributeAmount, 'Ether to account', i);
          distributeBlock = web3.eth.blockNumber;
        }
      }
      if (doesOneDistribute) console.log('');
      isDistributed = true;
    }
  }

  function addPeers(enodes) {
    if (!Array.isArray(enodes)) return;
    for (var i = 0; i < enodes.length; i++) {
      console.log('enodes', enodes[i]);
      web3.admin.addPeer(enodes[i]);
    }
  }

})();
